### PR TITLE
MISUV-6246 - pre-populated date on Date your business stopped page

### PR DESCRIPTION
**Pull request by JordHemsworth**

### CODE REVIEW

File: test/controllers/incomeSources/cease/IncomeSourceEndDateControllerSpec.scala [hunk 1]
```diff
@@ -451,4 +508,323 @@ class IncomeSourceEndDateControllerSpec extends TestSupport with MockAuthenticat
 
     }
   }
+
+  "Individual - IncomeSourceEndDateController.showChange" should {
```
Comment (SiddharthAtos) [inline test/controllers/incomeSources/cease/IncomeSourceEndDateControllerSpec.scala:512]: add the pre-filled date check test

File: app/controllers/incomeSources/cease/IncomeSourceEndDateController.scala [hunk 2]
```diff
@@ -214,4 +307,23 @@ class IncomeSourceEndDateController @Inject()(val authenticate: AuthenticationPr
...
+    if (isChange) {
+      sessionService.get(incomeSourceType.endDateSessionKey).flatMap {
+        case Right(Some(date)) =>
+          Future.successful(
+            form.fill(
+              DateFormElement(
+                LocalDate.parse(date)
+              )
+            ))
+        case _ => Future.failed(new Exception("[IncomeSourceEndDateController][getFilledForm]: Error getting endDateSessionKey:"))
```
Comment (SiddharthAtos) [inline app/controllers/incomeSources/cease/IncomeSourceEndDateController.scala:323]: change this "endDateSessionKey" in the log  to  ```${incomeSourceType.endDateSessionKey}```

File: it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala [hunk 3]
```diff
@@ -95,6 +101,62 @@ class IncomeSourceEndDateControllerISpec extends ComponentSpecBase {
...
+        When(s"I call GET $dateBusinessShowChangeUrl")
+        val result = IncomeTaxViewChangeFrontend.getBusinessEndDate
+        verifyIncomeSourceDetailsCall(testMtditid)
+
+        result should have(
+          httpStatus(OK),
+          pageTitleIndividual("incomeSources.cease.endDate.selfEmployment.heading"),
+          elementTextByID("income-source-end-date-hint")(hintText),
+          elementTextByID("continue-button")(continueButtonText)
+        )
```
Comment (SiddharthAtos) [inline it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala:120]: add pre-filled date check

File: it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala [hunk 4]
```diff
@@ -95,6 +101,62 @@ class IncomeSourceEndDateControllerISpec extends ComponentSpecBase {
...
 
+  s"calling GET $dateBusinessShowChangeUrl" should {
+    "render the Date Business Ceased Page" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with a business")
+        enable(IncomeSources)
+        IncomeTaxViewChangeStub.stubGetIncomeSourceDetailsResponse(testMtditid)(OK, businessOnlyResponse)
+
+        When(s"I call GET $dateBusinessShowChangeUrl")
+        val result = IncomeTaxViewChangeFrontend.getBusinessEndDate
```
Comment (SiddharthAtos) [inline it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala:112]: the request should be made to `change-business-end-date`

File: it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala [hunk 5]
```diff
@@ -149,6 +211,60 @@ class IncomeSourceEndDateControllerISpec extends ComponentSpecBase {
...
 
+  s"calling GET $dateUKPropertyShowChangeUrl" should {
+    "render the Date UK Property Ceased Page" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with UK property")
+        enable(IncomeSources)
+        IncomeTaxViewChangeStub.stubGetIncomeSourceDetailsResponse(testMtditid)(OK, ukPropertyOnlyResponse)
+
+        When(s"I call GET $dateUKPropertyShowChangeUrl")
+        val result = IncomeTaxViewChangeFrontend.getUKPropertyEndDate
```
Comment (SiddharthAtos) [inline it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala:222]: the request should be made to `change-uk-property-end-date`

File: it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala [hunk 6]
```diff
@@ -202,5 +318,59 @@ class IncomeSourceEndDateControllerISpec extends ComponentSpecBase {
...
+
+  s"calling GET $dateForeignPropertyShowChangeUrl" should {
+    "render the Date Foreign Property Ceased Page" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with Foreign property")
+        enable(IncomeSources)
+        IncomeTaxViewChangeStub.stubGetIncomeSourceDetailsResponse(testMtditid)(OK, foreignPropertyOnlyResponse)
+
+        When(s"I call GET $dateForeignPropertyShowChangeUrl")
+        val result = IncomeTaxViewChangeFrontend.getForeignPropertyEndDate
```
Comment (SiddharthAtos) [inline it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala:330]: the request should be made to `change-foreign-property-end-date`

File: it/controllers/agent/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala [hunk 7]
```diff
@@ -98,6 +104,63 @@ class IncomeSourceEndDateControllerISpec extends ComponentSpecBase {
...
+  s"calling GET $dateBusinessShowChangeAgentUrl" should {
+    "render the Date Business Ceased Page" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with a business")
+        enable(IncomeSources)
+        stubAuthorisedAgentUser(authorised = true)
+        IncomeTaxViewChangeStub.stubGetIncomeSourceDetailsResponse(testMtditid)(OK, businessOnlyResponse)
+
+        When(s"I call GET $dateBusinessShowChangeAgentUrl")
+        val result = IncomeTaxViewChangeFrontend.getBusinessEndDate(clientDetailsWithConfirmation)
```
Comment (SiddharthAtos) [inline it/controllers/agent/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala:116]: the request should be made to `change-business-end-date`

File: it/controllers/agent/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala [hunk 8]
```diff
@@ -155,6 +218,63 @@ class IncomeSourceEndDateControllerISpec extends ComponentSpecBase {
...
+  s"calling GET $dateUKPropertyShowChangeAgentUrl" should {
+    "render the Date UK Property Ceased Page" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with UK property")
+        enable(IncomeSources)
+        stubAuthorisedAgentUser(authorised = true)
+        IncomeTaxViewChangeStub.stubGetIncomeSourceDetailsResponse(testMtditid)(OK, ukPropertyOnlyResponse)
+
+        When(s"I call GET $dateUKPropertyShowChangeAgentUrl")
+        val result = IncomeTaxViewChangeFrontend.getUKPropertyEndDate(clientDetailsWithConfirmation)
```
Comment (SiddharthAtos) [inline it/controllers/agent/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala:230]: the request should be made to `change-uk-property-end-date`

File: it/controllers/agent/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala [hunk 9]
```diff
@@ -211,5 +331,62 @@ class IncomeSourceEndDateControllerISpec extends ComponentSpecBase {
...
+  s"calling GET $dateForeignPropertyShowChangeAgentUrl" should {
+    "render the Date Foreign Property Ceased Page" when {
+      "User is authorised" in {
+        Given("I wiremock stub a successful Income Source Details response with Foreign property")
+        enable(IncomeSources)
+        stubAuthorisedAgentUser(authorised = true)
+        IncomeTaxViewChangeStub.stubGetIncomeSourceDetailsResponse(testMtditid)(OK, foreignPropertyOnlyResponse)
+
+        When(s"I call GET $dateForeignPropertyShowChangeAgentUrl")
+        val result = IncomeTaxViewChangeFrontend.getForeignPropertyEndDate(clientDetailsWithConfirmation)
```
Comment (SiddharthAtos) [inline it/controllers/agent/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala:344]: the request should be made to `change-foreign-property-end-date`

File: app/controllers/incomeSources/cease/IncomeSourceEndDateController.scala [hunk 10]
```diff
@@ -165,47 +230,75 @@ class IncomeSourceEndDateController @Inject()(val authenticate: AuthenticationPr
...
+          )(user, messages)))
         },
 
         validatedInput => (incomeSourceTypeValue, id) match {
 
           case (SelfEmployment, None) =>
             Logger("application").error(s"${if (isAgent) "[Agent]"}" +
               s"[IncomeSourceEndDateController][handleSubmitRequest]: missing income source ID.")
-            errorHandler.showInternalServerError()
+            Future.failed(new Exception(s"[IncomeSourceEndDateController][handleSubmitRequest]: missing income source ID."))
```
[the remark this answers is not in the document]
Comment (JordHemsworth) [reply app/controllers/incomeSources/cease/IncomeSourceEndDateController.scala]: done

File: it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala (hunk 3 above)
Comment (JordHemsworth) [reply it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala:127]: added to UTs

File: it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala (hunk 6 above)
Comment (JordHemsworth) [reply it/controllers/incomeSources/cease/IncomeSourceEndDateControllerISpec.scala]: changed

**Review by SiddharthAtos — APPROVED** [review]

**Pull request merged**